File: typescript/src/reconnect.ts
import { reconnectDelayMs, type ReconnectBackoffPolicy } from './backoff.js';

export type ReconnectStatus = 'idle' | 'waiting' | 'connecting' | 'open' | 'stable';

export interface ReconnectState {
  /** Number used by the next exponential-delay calculation. */
  readonly attempt: number;
  readonly status: ReconnectStatus;
}

export interface ReconnectSchedule {
  readonly state: ReconnectState;
  readonly delayMs: number;
}

function validateAttempt(attempt: number): void {
  if (!Number.isSafeInteger(attempt) || attempt < 0) {
    throw new RangeError('attempt must be a non-negative safe integer');
  }
}

function reconnectState(attempt: number, status: ReconnectStatus): ReconnectState {
  validateAttempt(attempt);
  return Object.freeze({ attempt, status });
}

export function createReconnectState(): ReconnectState {
  return reconnectState(0, 'idle');
}

/** Calculates the next delay and advances the attempt before any timer or I/O runs. */
export function scheduleReconnectAttempt(
  state: ReconnectState,
  policy: ReconnectBackoffPolicy,
): ReconnectSchedule {
  validateAttempt(state.attempt);
  if (state.attempt === Number.MAX_SAFE_INTEGER) {
    throw new RangeError('reconnect attempt is exhausted');
  }
  return {
    delayMs: reconnectDelayMs(state.attempt, policy),
    state: reconnectState(state.attempt + 1, 'waiting'),
  };
}

export function markReconnectConnecting(state: ReconnectState): ReconnectState {
  return reconnectState(state.attempt, 'connecting');
}

/** An open transport is not yet considered stable and does not reset backoff. */
export function markReconnectOpen(state: ReconnectState): ReconnectState {
  return reconnectState(state.attempt, 'open');
}

export function markReconnectStable(state: ReconnectState): ReconnectState {
  validateAttempt(state.attempt);
  return reconnectState(0, 'stable');
}

export function ensureMinimumReconnectAttempt(
  state: ReconnectState,
  minimumAttempt: number,
): ReconnectState {
  validateAttempt(minimumAttempt);
  return reconnectState(Math.max(state.attempt, minimumAttempt), state.status);
}

export function resetReconnectState(): ReconnectState {
  return createReconnectState();
}

File: rust/src/reconnect.rs
use crate::{RealtimeError, ReconnectBackoffPolicy, reconnect_delay_ms};

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum ReconnectStatus {
    Idle,
    Waiting,
    Connecting,
    Open,
    Stable,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct ReconnectState {
    pub attempt: u32,
    pub status: ReconnectStatus,
}

impl Default for ReconnectState {
    fn default() -> Self {
        Self {
            attempt: 0,
            status: ReconnectStatus::Idle,
        }
    }
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct ReconnectSchedule {
    pub state: ReconnectState,
    pub delay_ms: u64,
}

/// Calculates a delay and advances the attempt before a timer or I/O runs.
///
/// # Errors
///
/// Returns policy errors from [`reconnect_delay_ms`] or
/// [`RealtimeError::ReconnectAttemptExhausted`] before the attempt wraps.
pub fn schedule_reconnect_attempt(
    state: ReconnectState,
    policy: ReconnectBackoffPolicy,
) -> Result<ReconnectSchedule, RealtimeError> {
    let attempt = state
        .attempt
        .checked_add(1)
        .ok_or(RealtimeError::ReconnectAttemptExhausted)?;
    Ok(ReconnectSchedule {
        delay_ms: reconnect_delay_ms(state.attempt, policy)?,
        state: ReconnectState {
            attempt,
            status: ReconnectStatus::Waiting,
        },
    })
}

#[must_use]
pub const fn mark_reconnect_connecting(state: ReconnectState) -> ReconnectState {
    ReconnectState {
        status: ReconnectStatus::Connecting,
        ..state
    }
}

#[must_use]
pub const fn mark_reconnect_open(state: ReconnectState) -> ReconnectState {
    ReconnectState {
        status: ReconnectStatus::Open,
        ..state
    }
}

#[must_use]
pub const fn mark_reconnect_stable(_state: ReconnectState) -> ReconnectState {
    ReconnectState {
        attempt: 0,
        status: ReconnectStatus::Stable,
    }
}

#[must_use]
pub const fn ensure_minimum_reconnect_attempt(
    state: ReconnectState,
    minimum_attempt: u32,
) -> ReconnectState {
    ReconnectState {
        attempt: if state.attempt > minimum_attempt {
            state.attempt
        } else {
            minimum_attempt
        },
        ..state
    }
}

#[must_use]
pub fn reset_reconnect_state() -> ReconnectState {
    ReconnectState::default()
}
